### ISSUE TITLE
feat(scripts/training): training script to run over all models, all dataset sizes, hyparparam optimize for final results

### ISSUE THREAD

**Issue by mlederbauer** [mlederbauer/NMRcraft#28]

**Comment by mlederbauer** [comment]
We want to plot the following results:

For each experiment outlined in Issue #24 :

- Run a Random Forest, SVC, Logistic Reg, Gradient Boosting
- Run each model with the dataset sizes 0.01, 0.1, 1 (how much of the dataset you sample)
- Get for each model the different scores

What do we want to get in the end?
For each baseline, 3-4 plots showing the ROC/F1-score/Accuracy like this
![image](https://github.com/mlederbauer/NMRcraft/assets/98785759/263712d2-7c03-4a8e-bb69-c722ef5f64a2)

We can beautify the plot later with LaTeX sans serif and the minecraft color scheme, but this would be the ~6*3 plots we need in the end.

The statement we want to get in the end:
"For the task of predicting the X1 ligand from the NMR tensors and metal, Model xyz performs best with the dataset sizes of xyz, we get an ROC of xxx +/- xxx (bootstrapping 10 times on test set)."

**Comment by mlederbauer** [comment]
Tagging @strsamue here, please let me know if you need more information

**Comment by strsamue** [comment]
Short Update: I am working on it but I have to generate a model class first with the training tuning and evaluation method. so it takes a bit. I did not push anything yet.

**Comment by mlederbauer** [comment]
Thank you for the update @strsamue ! Ah I see, so the models in `nmrcraft/models` don’t work for that purpose?

**Comment by strsamue** [comment]
no i was thinking more of something like this: 

```
from nmrcraft.model_configs import model_config
from nmrcraft.training.hyperparameter_tune import HyperparameterTuner
from nmrcraft.data.dataset import DataLoader
from nmrcraft.evaluation.evaluation import model_evaluation

class Classifier:
    def __init__(
        self,
        model_name="random_forest",
        max_evals=1
        target="metal"
        dataset_size = 0.01
        feature_columns = [
            "M_sigma11_ppm",
            "M_sigma22_ppm",
            "M_sigma33_ppm",
            "E_sigma11_ppm",
            "E_sigma22_ppm",
            "E_sigma33_ppm",
        ]
    ):
        model_config=model_configs[model_name]

        self.tuner = HyperparameterTuner(model_name, model_config, max_evals)
        
        self.X_train, self.X_test, self.y_train, self.y_test = DataLoader(feature_columns=feature_columns, target_columns=target, dataset_size=dataset_size).load_data()

    def hyperparameter_tune(
        self
    ):
        log.INFO(f"Performing Hyperparameter tuning for the Model ({self.model_name})")

        #DATA LEAKAGE!!! MUST be done by CV!!!!
        self.best_params, _ = self.tuner.tune(self.X_train, self.y_train, self.X_test, self.y_test)

    def train():
        model_func = lambda **params: load_model(
            model_name, **{**params, **self.model_config}
        )

        self.model = model_func(**self.best_params)
        best_model.fit(self.X_train, self.y_train)



    def evaluate():
        metrics, cm, fpr, tpr = model_evaluation(best_model, X_test, y_test)
        return metrics, cm, fpr, tpr
```



it's not finished but what do you think?

**Comment by mlederbauer** [comment]
Hi @strsamue ,super, thanks a lot for this! It’s super nice to have a class like this. Could you maybe start a draft pull request so that we can look in to it?

**Comment by strsamue** [comment]
I try to do it in the evening or tomorrow ;)

**Comment by mlederbauer** [comment]
Thank you; feel free to submit a “raw” draft pull request or publish the branch for us to take a look :)

**Comment by strsamue** [comment]
I just uploaded the first draft but I'm not finished at all feel free to add suggestions.  The branch is called: chore/#1-script-final-results.

**Comment by mlederbauer** [comment]
Thanks for the info on where to find the branch. Can you elaborate on the errors you received and where exactly you require help for finishing this part of the script? We are running out of time and really need this part of the script to work, to proceed and write down our results. We need this before we can finalize the report and the poster. Thank you for letting us know asap @strsamue .

**Comment by strsamue** [comment]
sorry completely forgot to upload the picture here you go:
<img width="742" alt="Screenshot 2024-05-09 at 22 32 59" src="https://github.com/mlederbauer/NMRcraft/assets/149368839/fb02c763-04a8-4a91-9f5e-721b6bea6509">

so if you get this to work then the roc should be done (might need to debug some small things), the rest of the plots can be added to the visualizer rather quickly I think, and then only formatting is needed

**Comment by mlederbauer** [comment]
@strsamue  Found the issue. `max_evals` (int=10)  was passed to the declaration of the Hyperparameter tuner, it needed to be changed to use the default value of `self.algo`. There was also another bug of passing the model_config to the Classifier class. The code runs for me now and is able to log the necessary results; the evaluation needs to be checked and cleaned though (I mostly commented it out for the time being, and had some issues installing kaleido via poetry/pip). In the end, we need to plot in the format for the poster and report (matplotlib pyplot).
Please check the draft PR #49.

**Comment by mlederbauer** [comment]
![image](https://github.com/mlederbauer/NMRcraft/assets/98785759/6362802c-6e81-406b-bf89-b0737a1da4e6)

**Comment by strsamue** [comment]
Perfect thanks for finding the error! I'll have a look at the evaluation ;)

**Comment by strsamue** [comment]
I got the bootstrapping working only need to add the errorbars😉 maybe i got time tonight or tomorrow😉

**Comment by mlederbauer** [comment]
Amazinggg! Thanks so much 👍👍

**Comment by strsamue** [comment]
Cose still needs some cleaning up but it works. We get this out. Did not run it yet with lots of tuning etc. will do that at home.
![accuracy](https://github.com/mlederbauer/NMRcraft/assets/149368839/00d92717-d445-4cf9-9aed-f365ecc994c3)
![f1_score](https://github.com/mlederbauer/NMRcraft/assets/149368839/eca9e0da-53a9-4a36-80d4-8af9929c6d42)
![ROC_Plot_logistic_regression](https://github.com/mlederbauer/NMRcraft/assets/149368839/22feaffe-c774-46c3-8536-96522a6cf21e)

Pass das so?


PS: ROC bekommen wir für jeden classifier

**Comment by mlederbauer** [comment]
Super super nice, vielen Dank dir! Should we open a PR to review ?

**Comment by mlederbauer** [comment]
Which target is this by chance? With the confusion matrix, we had the issue that the targets X1, X2, X3, X4 didn’t work as only chloride was chosen all the time (“dummy classifier”)

**Comment by strsamue** [comment]
this is just metal but I coded it to be universal it should work with all hopefully I can check it tomorrow or friday.
